### ISSUE TITLE
install error on Solaris

### ISSUE THREAD

**Issue by lhoward** [Azure/azure-storage-blob-go#96]
Similar issue to #23 on Solaris. Worked around by creating `azblob/zc_mmf_solaris.go` with following contents (from some searching/copy/pasting – I don't know Go).

```go
// +build solaris

package azblob

import (
        "os"
        "syscall"
        "golang.org/x/sys/unix"
)

type mmf []byte

func newMMF(file *os.File, writable bool, offset int64, length int) (mmf, error) {
        prot, flags := unix.PROT_READ, syscall.MAP_SHARED // Assume read-only
        if writable {
                prot, flags = syscall.PROT_READ|syscall.PROT_WRITE, syscall.MAP_SHARED
        }
        addr, err := unix.Mmap(int(file.Fd()), offset, length, prot, flags)
        return mmf(addr), err
}

func (m *mmf) unmap() {
        err := unix.Munmap(*m)
        *m = nil
        if err != nil {
                panic("if we are unable to unmap the memory-mapped file, there is serious concern for memory corruption")
        }
}
```

**Comment by zezha-msft** [comment]
Hi @lhoward, thanks for reaching out!

I've logged this item to provide support for solaris.

**Comment by lhoward** [comment]
Re-reading that patch, and with about 1% more Go knowledge than when I filed the report, it should not mix the syscall and unix packages (so, remove import syscall and s/syscall/unix/).

**Comment by petems** [comment]
I think we can just add the solaris build tag to the existing file right?

Opened a PR to do so 👍

**Comment by lhoward** [comment]
No, it's a new file that uses `golang.org/x/sys/unix` instead of the `syscall` package; at least on the version of Go I have, syscall.Mmap and syscall.Munmap are unavailable on Solaris. The version below avoids mixing the two packages.

```go
// +build solaris

package azblob

import (
        "os"
        "golang.org/x/sys/unix"
)

type mmf []byte

func newMMF(file *os.File, writable bool, offset int64, length int) (mmf, error) {
        prot, flags := unix.PROT_READ, unix.MAP_SHARED // Assume read-only
        if writable {
                prot, flags = unix.PROT_READ|unix.PROT_WRITE, unix.MAP_SHARED
        }
        addr, err := unix.Mmap(int(file.Fd()), offset, length, prot, flags)
        return mmf(addr), err
}

func (m *mmf) unmap() {
        err := unix.Munmap(*m)
        *m = nil
        if err != nil {
                panic("if we are unable to unmap the memory-mapped file, there is serious concern for memory corruption")
        }
}
```

**Comment by lhoward** [comment]
(Above goes in zc_mmf_solaris.go.)

**Comment by petems** [comment]
Yep, I just realised that after the build failed. IIRC syscall is now frozen so switching to the `"golang.org/x/sys/unix"` makes sesnse

**Comment by jclulow** [comment]
I just hit this on an illumos system while trying to build Hugo, the static site generator.  In the upcoming Go 1.13, there will be an `illumos` build tag.  It'll imply the `solaris` build tag, though, so fixing this for `solaris` will fix it for both.

Does it make sense to just use `sys/unix` instead of `syscall` for _all_ UNIX platforms?  Can I help somehow with testing (e.g., on illumos systems) to get this over the line?  Thanks.

**Comment by zezha-msft** [comment]
Support for Solaris has been released in v0.10.0.

**Comment by jclulow** [comment]
Thanks!
